### PR TITLE
Adding MeshBase::bid_nodes_{begin,end}

### PR DESCRIPTION
**Pull request by jwpeterson**
This allows for iteration over nodes with a user-selected boundary id.

@andrsd

### CODE REVIEW

**Comment by permcody** [comment]
Cool, Are there any places where we might need to take advantage of this in
MooseMesh or BoundaryRestrictable?

**Comment by andrsd** [comment]
@jwpeterson Nice!
@permcody My intended use was for transfers. Nearest-node transfer and projection transfers will benefit from this (maybe others as well)

**Comment by jwpeterson** [comment]
On Thu, Mar 26, 2015 at 10:40 AM, Cody Permann <notifications@github.com>
wrote:

> Cool, Are there any places where we might need to take advantage of this in
> MooseMesh or BoundaryRestrictable?
>

Looks like right now we just use the BoundaryInfo object
in MooseMesh::buildNodeList() to create our vector of BndNode* and BndElem*
objects ahead of time.  I don't know about BoundaryRestrictable, this was
mostly for David's transfer thing as he mentioned.

-- 
John

**Comment by permcody** [comment]
Yes, I knew what it was for. Just wanted to see if there was potential for
clean up anywhere else as a result

**Comment by aeslaughter** [comment]
I don't think this will be needed in the restrictable interfaces, since
there is no looping. But, I think some of the MeshModifiers could benefit
from this.

**Comment by jwpeterson** [comment]
It should be relatively easy to add one more type of iterator that loops over *all* boundary nodes rather than just the ones with a specified ID, so I'm probably going to tack that onto this PR as well...

File: include/mesh/mesh_base.h [hunk 1]
```diff
@@ -852,6 +854,8 @@ class MeshBase : public ParallelObject
   virtual const_node_iterator local_nodes_end    () const = 0;
   virtual const_node_iterator pid_nodes_begin    (const processor_id_type proc_id) const = 0;
   virtual const_node_iterator pid_nodes_end      (const processor_id_type proc_id) const = 0;
+  virtual const_node_iterator bid_nodes_begin    (const boundary_id_type bndry_id) const = 0;
```
Comment (roystgnr) [inline include/mesh/mesh_base.h]: This is probably a good place to break our long-standing "don't properly document iterator range methods" tradition.  I'd have assumed that the behavior here was "iterate through nodes that are on sides with the given boundary id", but that only happens if the nodes themselves also share that id, right?  We should probably clarify that with a comment.

Comment (jwpeterson) [inline include/mesh/mesh_base.h]: It only iterates through nodes for which 

I agree that all these iterator methods should probably be documented.  I was thinking about reorganizing them so the const and non-const versions are together and we can put one 
```
/**
*
*/
```
section that covers all 4.

**Comment by roystgnr** [comment]
Looks good, thanks!

**Pull request merged**